client/db/src/kv/mod.rs

```diff
@@ -0,0 +1,359 @@
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+// This file is part of Frontier.
+//
+// Copyright (c) 2021-2022 Parity Technologies (UK) Ltd.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#[cfg(feature = "parity-db")]
+mod parity_db_adapter;
+mod upgrade;
+mod utils;
+
+use std::{
+	marker::PhantomData,
+	path::{Path, PathBuf},
+	sync::Arc,
+};
+
+use parking_lot::Mutex;
+use scale_codec::{Decode, Encode};
+// Substrate
+pub use sc_client_db::DatabaseSource;
+use sp_blockchain::HeaderBackend;
+use sp_core::H256;
+pub use sp_database::Database;
+use sp_runtime::traits::Block as BlockT;
+// Frontier
+use crate::TransactionMetadata;
+use fp_storage::{EthereumStorageSchema, PALLET_ETHEREUM_SCHEMA_CACHE};
+
+const DB_HASH_LEN: usize = 32;
+/// Hash type that this backend uses for the database.
+pub type DbHash = [u8; DB_HASH_LEN];
+
+/// Database settings.
+pub struct DatabaseSettings {
+	/// Where to find the database.
+	pub source: DatabaseSource,
+}
+
+pub(crate) mod columns {
+	pub const NUM_COLUMNS: u32 = 4;
+
+	pub const META: u32 = 0;
+	pub const BLOCK_MAPPING: u32 = 1;
+	pub const TRANSACTION_MAPPING: u32 = 2;
+	pub const SYNCED_MAPPING: u32 = 3;
+}
+
+pub mod static_keys {
+	pub const CURRENT_SYNCING_TIPS: &[u8] = b"CURRENT_SYNCING_TIPS";
+}
+
+#[derive(Clone)]
+pub struct Backend<Block: BlockT> {
+	meta: Arc<MetaDb<Block>>,
+	mapping: Arc<MappingDb<Block>>,
+}
+
+#[async_trait::async_trait]
+impl<Block: BlockT> crate::BackendReader<Block> for Backend<Block> {
+	async fn block_hash(
+		&self,
+		ethereum_block_hash: &H256,
+	) -> Result<Option<Vec<Block::Hash>>, String> {
+		self.mapping().block_hash(ethereum_block_hash)
+	}
+	async fn transaction_metadata(
+		&self,
+		ethereum_transaction_hash: &H256,
+	) -> Result<Vec<TransactionMetadata<Block>>, String> {
+		self.mapping()
+			.transaction_metadata(ethereum_transaction_hash)
+	}
+	async fn filter_logs(
+		&self,
+		_from_block: u64,
+		_to_block: u64,
+		_addresses: Vec<sp_core::H160>,
+		_topics: Vec<Vec<Option<H256>>>,
+	) -> Result<Vec<crate::FilteredLog>, String> {
+		// KeyValue db does not index logs
+		unimplemented!();
```

Comment:
Will this cause potential panic for the kv backend client? Better to return empty?

Reply:
This is already checked if the backend is indexed or not [here ](https://github.com/paritytech/frontier/pull/883/files/092bad72db0eafacf3b37ff7259d35d394e5c059#diff-33da185d821c2df8279916184aea8664aba5bcf80e76ce9e02508ce921ae54c2R266) so will not cause a panic. This ensures that the indexed logs are only returned for supported backends while the unsupported ones use `filter_logs` instead.

Reply:
Got it.

Reply:
That can be a possibility to return `Err("not supported")`